### PR TITLE
Smart Sampling for Monitoring

### PR DESCRIPTION
**Pull request by imandrealombardo**
### Description
Implemented a pipeline step to smart sample images from a database of client images. This is to re-train and evaluate the model.

### Related Jira Ticket
- Ticket: [BCV-52](https://gemeente-amsterdam.atlassian.net/browse/BCV-52)
- Description: Create script to perform the following features:
    - Randomly sample 10 images from each batch of images
    - Store the images in a folder 
    - Classify images based on detection counts and put in different bins
    - For each bin, randomly sample images for retraining
    - Store images for retraining in daytime folder which matches the starting time of the pre-inference pipeline

### CODE REVIEW

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 1]
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
...
+    detection_range : int
+        The range of detection counts across all images.
+
+    Returns
+    -------
+    int
+        The number of bins to be used for categorization.
+    """
+    
+    if detection_range <= 10:
```
[the remark this answers is not in the document]
Comment (sdavrieux) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:200]: And the number of bins?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 2]
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
...
+        return 3
+    elif 10 < detection_range <= 50:
+        return 5
+    else:
+        return 10
+
+def initialize_bin_counts(bins: np.ndarray) -> Tuple[Dict[str, List], List[str]]:
+    """
+    Initializes the counts and labels for bins.
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:209]: Can you explain better what this function does?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 3]
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
...
+    Categorizes each image into a bin based on its detection count.
+
+    Parameters
+    ----------
+    image_counts : Dict[str, int]
+        A dictionary mapping image identifiers to their detection counts.
+    bins : np.ndarray
+        The array of bin edges.
+    bin_labels : List[str]
+        A list of bin labels.
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:241]: What are the bin labels? Can you make me an example?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 4]
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
...
+    -------
+    None
+        The function does not return anything. It performs the operation of copying the sampled images.
+    """
+    
+    for upload_date, images in sampled_images_by_date.items():
+        # Copy the sampled images
+        for image in images:
+            formatted_upload_date = upload_date.strftime("%Y-%m-%d_%H_%M_%S")
+            image_filename = image[2]  # Assuming image is a tuple (customer_name, upload_date, image_name)
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:317]: All these assumptions of this data type you're using make the code a bit confusing for a person who tries to read it for the first time without any knowledge.
I would suggest to start creating your own data type to represent an Image, and then we can see how to simplify everything a bit further.

Comment (sdavrieux) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:317]: This data object could contain the details of the image you're using and for example also the number of detections and stuff like that.

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 5]
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
...
+    Raises
+    ------
+    ValueError
+        If database credentials are not provided.
+    """
     database_parameters = json.loads(database_parameters_json)
-    db_username = database_parameters["db_username"]
-    db_name = database_parameters["db_name"]
-    db_hostname = database_parameters["db_hostname"]
+    db_username = database_parameters.get("db_username")
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:476]: Why get instead of []?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 6]
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
...
-            random_values = random.sample(values, 10)
+        if len(values) >= n_images_to_sample:
+            random_values = random.sample(values, n_images_to_sample)
         else:
             random_values = values
         random_result[key] = random_values
 
     return random_result
 
+def connect_to_database(database_parameters_json):
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:456]: What is the reason to split the connect_to_database?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 7]
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
...
+                            images_statistics[upload_date] = extracted_data
+                            
+    except SQLAlchemyError as e:
+        logger.error(f"Database operation failed: {e}")
+    except ValueError as e:
+        logger.error(f"Configuration error: {e}")
+    except Exception as e:
+        logger.error(f"An unexpected error occurred: {e}")
+
+    return images_statistics, image_counts
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:564]: The image counts can be counted from image statistics no?
What if instead of using all these multiples arrays with information that makes the code hard to follow we just use a dataframe with the actual data and we just share that across the functions?

File: blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py [hunk 8]
```diff
@@ -64,6 +76,7 @@ def cleaning_pipeline():
     # Retrieve values from the YAML
     BlurringAsAServiceSettings.set_from_yaml("config.yml")
     settings = BlurringAsAServiceSettings.get_settings()
+    sampling_settings = settings["sampling_parameters"]
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py:79]: Where is this used?

File: blurring_as_a_service/settings/settings_schema.py [hunk 9]
```diff
@@ -111,6 +111,12 @@ class InferenceCustomerPipelineSpec(SettingsSpecModel):
     model_name: str
     model_version: str
     model_parameters: InferenceModelParameters
+    
+
+class SmartSamplingPipelineSpec(SettingsSpecModel):
```
Comment (sdavrieux) [inline blurring_as_a_service/settings/settings_schema.py:116]: Good job :wink:

File: config.example.yml [hunk 10]
```diff
@@ -85,4 +85,4 @@ logging:
     level: INFO
     format: "%(asctime)s|||%(levelname)-8s|%(name)s|%(message)s"
     datefmt: "%Y-%m-%d %H:%M:%S"
-  ai_instrumentation_key: "APPLICATION_INSIGHTS_INSTRUMENTATION_KEY"
\ No newline at end of file
+  ai_instrumentation_key: "APPLICATION_INSIGHTS_CONNECTION_STRING"
```
Comment (sdavrieux) [inline config.example.yml:88]: Missing end of the file newline.

Comment (epureanudiana) [reply config.example.yml:88]: `"smart_sampling"` should be added to the list: 

```
own_packages: ["__main__", "blurring_as_a_service", "yolov5", "evaluate_with_cvt_metrics", "validate_model",
                 "evaluate_with_coco", "detect_and_blur_sensitive_data", "move_files", "split_workload"]
```
                 
did it work without it when you tested it?

File: blurring_as_a_service/utils/generics.py [hunk 11]
```diff
@@ -40,8 +40,10 @@ def delete_file(file_path):
 
 
 def copy_file(relative_path, input_path, output_path):
-    source_path = os.path.join(input_path, relative_path)
-    destination_path = os.path.join(output_path, relative_path)
+
+    # We don't use os.path.join because it doesn't work with Azure paths
+    source_path = input_path + relative_path
```
Comment (epureanudiana) [inline blurring_as_a_service/utils/generics.py:45]: how did it work before then?

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py (hunk 4 above)
Comment (epureanudiana) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:317]: I agree! There are a lot of functions and it's difficult to see how they all come together :)

Firstly, I suggest you move all of these functions outside of the `cleaning_pipeline/components/smart_sampling.py` and put them under `cleaning_pipeline/source/sampling_pipeline.py`. We usually have the `components/*.py `to only contain minimal code.

This way the components part remains clean, something like this: 

 ```
 @command_component(name="smart_sampling")
 def smart_sampling(
     input_structured_folder: Input(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_cvt_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_quality_check_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_retraining_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     database_parameters_json: str,
     customer_name: str,
 ):
 
 SmartSampling.sample_images_for_quality_check(...)
 SmartSampling.sample_images_for_retraining(...)
 
 ```

You can create a class for sampling; in the `__init__` you can set attributes that you are now loosely collecting, for example `sample_images_for_retraining`, `n_images_to_sample` and you can also call the rest of the functions in there, whatever you prefer.

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (imandrealombardo) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:200]: These numbers are dependent on the test images results: on 2000 images with a number of detections per image ranging from 0 to 242, 10 bins allowed for a good distribution of the images.

File: blurring_as_a_service/utils/generics.py (hunk 11 above)
Comment (imandrealombardo) [reply blurring_as_a_service/utils/generics.py:45]: I don't know, but I was stuck for days because using `os.path.join()` didn't create the correct path (it was ignoring `input_path` and `output_path`). The only way I found to make it work was to concatenate the two paths with a `+`!

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py (hunk 3 above)
Comment (imandrealombardo) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:250]: Each bin label is a `string` descriptor of the range of detection counts that belongs to each respective bin. When images are processed and their detection counts are categorized, they are placed into these bins based on where their count falls in these ranges. 

For example, if we have a range of detection counts from 0 to 100, we decide to create bins like [0-9], [10-19], ...., [90-99], [100]. The text representing each range (like "0-9" or "10-19") is a bin label. In this example, an image with 15 detections would fall into the bin labeled "10-19".

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py (hunk 6 above)
Comment (imandrealombardo) [reply blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:465]: The purpose was to make the database connection code reusable throughout the code instead of repeating it and to enhance readibility. In this case we only connect to the database once, so it may not be needed after all

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 12]
```diff
@@ -96,300 +87,21 @@ def smart_sampling(
...
-    )
-    
-    # Collect images above the confidence score threshold from the database
-    # The threshold is set in config.yml as conf_score_threshold
-    conf_score_threshold = sampling_parameters["conf_score_threshold"]
-    _, image_counts = collect_images_above_threshold_from_db(
-        database_parameters_json, grouped_images_by_date, customer_name, conf_score_threshold
-    )
+    # Define the SmartSampling object
+    smartSampling = SmartSampling(input_structured_folder, customer_quality_check_folder, customer_retraining_folder, database_parameters_json,
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:97]: We use underscores and not camelCase 😉

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+from yolov5.baas_utils.database_handler import DBConfigSQLAlchemy  # noqa: E402
+from yolov5.baas_utils.database_tables import DetectionInformation  # noqa: E402
+
+from blurring_as_a_service.utils.generics import (  # noqa: E402
+    copy_file,
+)
+
+logger = logging.getLogger(__name__)
+
+class SmartSampling:
```
[the remark this answers is not in the document]
Comment (sdavrieux) [reply blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:32]: This is a SmartSampler not the SmartSampling

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+        The destination folder path where sampled images for retraining are stored.
+    database_parameters_json : str
+        JSON string with database connection parameters.
+    customer_name : str
+        The name of the customer associated with the images.
+    sampling_parameters : dict
+        A dictionary containing parameters for the sampling process, such as sample size, confidence score 
+        threshold, and sampling ratio.
+
+    Methods
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:54]: We can remove the methods from here. 
Examples: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

File: blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py [hunk 15]
```diff
@@ -3,36 +3,47 @@
...
+
 from blurring_as_a_service.settings.settings import (  # noqa: E402
     BlurringAsAServiceSettings,
 )
-from blurring_as_a_service.utils.generics import (  # noqa: E402
-    copy_file,
-    find_image_paths,
+from blurring_as_a_service.cleaning_pipeline.source.sampling_pipeline import (  # noqa: E402
+    SmartSampling,
+)
```
Comment (epureanudiana) [inline blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:23]: this import should come after line 39 where we setup the logger

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+    def __init__(self, input_structured_folder, customer_quality_check_folder, customer_retraining_folder, 
+                 database_parameters_json, customer_name, sampling_parameters):
+        self.input_structured_folder = input_structured_folder
+        self.customer_quality_check_folder = customer_quality_check_folder
+        self.customer_retraining_folder = customer_retraining_folder
+        self.database_parameters_json = database_parameters_json
+        self.customer_name = customer_name
+        self.sampling_parameters = sampling_parameters
+        
+    def sample_images_for_quality_check(self, grouped_images_by_date: Dict[str, List[str]]
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: All methods that are "private" should be underscored.

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+        None
+        """
+        
+        # Sample a number of random images for manual quality check
+        # The number is set in config.yml as quality_check_sample_size
+        n_images_to_sample = self.sampling_parameters["quality_check_sample_size"]
+        
+        logger.info(f'Sampling {n_images_to_sample} images for quality check.. \n')
+        
+        quality_check_images = SmartSampling.get_n_random_images_per_date(grouped_images_by_date, n_images_to_sample)
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: If a method is static you need to add the decorator and remove the self.
If the method is not static then it should be called by the same object, so self.method().
Otherwise you are calling it as a class method.

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+        grouped_images_by_date : Dict[str, List[str]]
+            A dictionary mapping dates (in 'YYYY-MM-DD_HH_MM_SS' string format) to lists of image filenames. 
+            These filenames are used to query the database and collect detection data for the corresponding images.
+            
+        Returns
+        -------
+        None
+        """
+        
+        # Collect images above the confidence score threshold from the database
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:156]: Comment is redundant due to good naming on function and variables.
I'm a big fan (even if some people don't agree) of a no comments between the code, my theory is that if you need to write a comment then it means that that code needs to be refactored on a better structure.

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+                            for data in extracted_data:
+                                data['count'] = count
+                                
+                            if upload_date in images_statistics:
+                                images_statistics[upload_date].extend(extracted_data)
+                            else:
+                                images_statistics[upload_date] = extracted_data
+                                
+                # Flatten the dictionary to a list of dictionaries
+                flat_list = [item for sublist in images_statistics.values() for item in sublist]
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: This is an example of something that needs a comment and therefore should be refactored into a function for clarity. Possibly in a utils file where if someone needs to redo this in the future can just call the same function.

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+            
+            # Ensure at least one image is sampled if total_images_to_sample > 0
+            total_images_to_sample = max(total_images_to_sample, 1) if total_images > 0 else 0
+            
+            logger.info(f'Total images to sample on date {upload_date}: {total_images_to_sample}')
+            
+            # Check if 'bin_label' column exists
+            if 'bin_label' not in df_date.columns:
+                logger.error("bin_label column not found in DataFrame for date: " + str(upload_date))
+                continue  # Skip to next date or handle error as needed
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: Not a big fan of continues and breaks, there is a lot of discussion if they are bad practice or not. 
I found easier to follow the flow of the program avoiding them.
@epureanudiana what is your idea about it?

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+
+        Returns
+        -------
+        int
+            The number of bins to be used for categorization.
+        """
+        
+        if detection_range <= 10:
+            return 3
+        elif 10 < detection_range <= 50:
```
Comment (sdavrieux) [inline blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: else is irrelevant since there is a return, not an error but just a style thing :P
Also you could do:
```
if detection_range > 50:
   return 10
if detection_range > 10:
   return 5
return 3
```

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+            images_per_bin = total_images_to_sample // len(bin_labels)
+            remainder = total_images_to_sample % len(bin_labels)
+            logger.info(f'Images per bin: {images_per_bin}, Remainder: {remainder}')
+            
+            for bin_label in bin_labels:
+                df_bin = df_date[df_date['bin_label'] == bin_label]
+                
+                # Adjust the number of images to sample from this bin
+                num_to_sample = min(images_per_bin + (1 if remainder > 0 else 0), len(df_bin))
+                remainder -= 1 if remainder > 0 else 0
```
[the remark this answers is not in the document]
Comment (imandrealombardo) [reply blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: I made a function out of this to explain it

File: blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py [hunk 23]
```diff
@@ -33,13 +33,25 @@ def cleaning_pipeline():
...
+        mode="rw_mount",
+        path=customer_quality_check_path,
+    )
+
+    customer_retraining_path = aml_interface.get_datastore_full_path(
+        f"{customer_name}_retraining"
     )
-    smart_sampling_step.outputs.customer_cvt_folder = Output(
+
+    smart_sampling_step.outputs.customer_retraining_folder = Output(
```
Comment (epureanudiana) [inline blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py:51]: can you please add a description for the Inputs and Outputs objects? also the ones that are commented from the deletion phase,

like this:
```
 smart_sampling_step.outputs.customer_quality_check_folder = Output(
         type=AssetTypes.URI_FOLDER,
         mode="rw_mount",
         path=customer_quality_check_path,
         description="a short description"
```

File: blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py (hunk 20 above)
Comment (epureanudiana) [reply blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py]: yea, previously we replaced it with if else

**Review by sdavrieux — APPROVED** [review]
Awesome job! Congrats on your first PR!
